### PR TITLE
[fs-detectors] process detectFramework in parallel

### PR DESCRIPTION
**Pull request by pbt**
The Vercel API's `detect-framework` API contacts GitHub's API using its own file adapter. 

However, currently the `detectFramework` function in the Vercel CLI (on which the API depends) calls the file system adapter in series using a `for` loop. This results in large amounts of lag when in a networked situation such as calling GitHub's API as each API call takes a couple hundred milliseconds.

I propose changing `detectFramework` to process the directories it scans in parallel.

Linked to: https://github.com/vercel/turbo-internal/issues/193

#### Tests

- [x] The code changed/added as part of this PR has been covered with tests
- [x] All tests pass locally with `yarn test-unit`

#### Code Review

- [x] This PR has a concise title and thorough description useful to a reviewer
- [x] Issue from task tracker has a link to this PR

### CODE REVIEW

**Review by agadzik — APPROVED** [review]
This looks good to me - I'll defer to the owners to see if theres anything odd about this PR!

File: packages/fs-detectors/src/detect-framework.ts [hunk 1]
```diff
@@ -76,15 +77,37 @@ async function matches(fs: DetectorFilesystem, framework: BaseFramework) {
...
+  try {
+    framework = await Promise.any(
+      frameworkList.map(async framework => {
+        if (await matches(fs, framework)) {
+          return framework.slug;
+        }
+        throw new Error(FRAMEWORK_ERR);
+      })
+    );
+  } catch (errorOrErrors: any) {
```
Comment (styfle) [inline packages/fs-detectors/src/detect-framework.ts]:
```suggestion
  } catch (errorOrErrors: unknown) {
```

**Review by styfle — DISMISSED** [review]
This will likely cause bugs because the order of the match matters and the order of the promise resolution is no longer guaranteed with Promise.any(). For example, the user might have both `next` and `storybook` in their package.json `dependencies`. Since `next` is the first in the array, it should match first, regardless of other matches (there can only be one framework selected afterall).

**Comment by agadzik** [comment]
> This will likely cause bugs because the order of the match matters. For example, the user might have `next` and they might have `storybook` in their package.json `dependencies`, however since `next` is the first in the array, it should match.

🤔 Maybe we switch to `Promise.all` if thats the case and then return back the first successful result...The con of this approach is that we make more network calls (one for each framework) that will 404, but it speeds up the overall method.

**Comment by pbt** [comment]
Will benchmark on `vercel/api` and report back on the results in a few hours.

**Review by styfle — APPROVED** [review]
Thanks, looks great! 🥇

**Pull request merged**